### ISSUE TITLE
Not all Endpoints take csv format request

### ISSUE THREAD

**Issue by VEnvirons** [iexg/IEX-API#95]
CSV is much more fast, simple, cpu and memory friendly.  As such, I'm not sure why it is treated as a "second class citizen" so to speak.  While it is used with some query endpoints, it is not availible across the board and unfortunately there are some endpoints that ONLY return JSON?  As a programmer in C++ and C# csv is preferred, assuming that if the field is empty, it is still included keeping the same number of fields per csv line, i.e.,  ",,"
This link explains and compares the two...

https://syslog.ravelin.com/lemony-scale-its-a-series-of-unfortunate-decisions-b16a59833146

**Comment by joshuablackburn-iex** [comment]
@VEnvirons You are correct that CSV is not supported for every endpoint, but we have made it available for most.  As developers of a high performance stock exchange, we understand the need for efficient data structures, and would not consider CSV a "second class citizen".  Our goal is to provide an easy to use, comprehensive API that can be used by everyone, for free.  Additionally, we want to provide features rapidly, and JSON is currently the common language of REST APIs.  The platform is actively developed and will continue to improve over time.  

Is there is a particular endpoint you're having an issue with that can be prioritized to support CSV?

**Comment by VEnvirons** [comment]
Of cours IEX is tremendous, no disagreement there.

However, IMHO, string manipulation, for example splitting a string by a comma (any) delimiter is far more common and popular, across languages and is typically part of a languages string methods.  There is no need for additional libraries or classes to be imported for dealing with json for example.  Also, csv uses less bandwidth, less CPU load and cycles, unless of course the cpu load or cycles increases on your end to make the csv?

As far as endpoint goes, I'd have to take a look again and make notes.

-S

**Comment by hvolentine** [comment]
I would like the Batch Quote endpoint to support .csv files.

**Comment by hvolentine** [comment]
Is there an IEX endpoint that will allow me to retrieve market indices such as DJIA Dow Jones Industrial Averages and S&P500?  If so, please reply with an example of the API link.

**Comment by joshuablackburn-iex** [comment]
@hvolentine Batch quote csv support may be a bit more difficult to implement at this time due to the variable nature of the returned data structure.  

Additionally, we do not provide the actual indices, though you can use their ETF's which track closely.  The actual index price requires a data contract with the index provider for each user, and is cost prohibitive.

**Comment by hvolentine** [comment]
Thank for the reply to the Batch Quote .csv and indices questions.  
I currently retrieve the Batcg Quote JSON file and parse it into a .CSV file with RegEx.  This works well but, a .csv feed would be easier and eliminate the RegEx processing steps.

**Comment by VEnvirons** [comment]
@Josh  I tend to agree with hvolentine (as I stated before) that csv is easier to parse and should be a format option for all endpoints.

With respect to "variable nature of returned data structure" is concerned.  I would say only that it is a text/ascii and as long as there are no commas contained in any data values (if it does, remove them) then it is really a matter of 1.) the number of fields and 2.) empty return values

For example, if an end point, for what ever reason didn't have any data, but was suppose to return date,open,high,low,close, then the return csv should look like ",,,," or if only the close was availible it would look like ",,,24.25"

Josh, do you have an example of the "Varible format" you mentioned"?

Bottom line is that end user4es can parse just about anything and csv is just another great format.

**Comment by jcape** [comment]
@VEnvirons stock/market/batch is a single endpoint, which returns a minimum of a three-dimensional data set. https://api.iextrading.com/1.0/stock/market/batch?symbols=goog,aapl&types=book for example, returns a 5-dimensional dataset.

CSV is capable of naively representing a two-dimensional dataset. It's not impossible to represent data with more than two dimensions, but you need to decide, describe, implement, and test how extra dimensions are going to be folded.

This is on our list as an enhancement, so I'm going to close discussion on this until we have something else to report on the issue.
